gcp/gcp.go

```diff
@@ -1127,3 +1144,109 @@ func getDatabaseInstanceStatus(instance *sqladmin.DatabaseInstance) types.State
+}
+
+func getClusters(listClusters *dataproc.ClusterIterator, region string) ([]*types.Cluster, error) {
+	clusters := make([]*types.Cluster, 0)
+	for {
+		cluster, err := listClusters.Next()
+		if err != nil && err == iterator.Done {
+			break
+		} else if err != nil {
+			log.Debugf("[GET_CLUSTERS] Error fetching cluster list: %+v", err)
+			return nil, err
+		}
+		clusters = append(clusters, newCluster(cluster, region))
+	}
+	return clusters, nil
+}
+
+func newCluster(cluster *dataprocpb.Cluster, region string) *types.Cluster {
+	log.Debugf("[GET_CLUSTERS] %+s in %s has status: %+s", cluster.ClusterName, region, cluster.Status.State)
+	return &types.Cluster{
+		Name:      cluster.ClusterName,
+		Uuid:      cluster.ClusterUuid,
+		Created:   cluster.Status.StateStartTime.AsTime(),
+		CloudType: types.GCP,
+		Region:    region,
+		Tags:      convertTags(cluster.Labels),
+		Config:    cluster.GetConfig(),
+		State:     getClusterState(cluster.Status.GetState()),
+	}
+}
+
+func getClusterState(s dataprocpb.ClusterStatus_State) types.State {
+	// convert ClusterSTatus_State to types.State
+
+	// This array must match the const ordering from here:
+	// https://pkg.go.dev/google.golang.org/genproto/googleapis/cloud/dataproc/v1#ClusterStatus_State
+	statuses := []types.State{
+		types.Unknown,
+		types.Creating,
+		types.Running,
+		types.Error,
+		types.Deleting,
+		types.Updating,
+		types.Stopping,
+		types.Stopped,
+		types.Starting,
+	}
+
+	for i, status := range statuses {
+		if i == int(s) {
+			return status
+		}
+	}
+	return types.Unknown
+}
+
+func (p gcpProvider) GetClusters() ([]*types.Cluster, error) {
+	log.Info("[GET_CLUSTERS] Fetching clusters across all regions. (slow)")
+	var clusters []*types.Cluster
+
+	ctx := context.Background()
+	req := &dataprocpb.ListClustersRequest{
+		ProjectId: p.projectID,
+		Region:    "global",
+	}
+
+	// scan global region for clusters
+	clist, err := getClusters(p.dataprocClient.ListClusters(ctx, req), "global")
+	if err != nil {
+		log.Errorf("[GET_CLUSTERS] Error listing clusters: %+v", err)
+		return nil, err
+	}
+	clusters = clist
+
+	// scan each geographic region for clusters
+	regionsList, err := getRegions(p)
+	if err != nil {
+		log.Errorf("[GET_CLUSTERS] Error fetching regions: %s", err)
+		return nil, err
+	}
+	for _, r := range regionsList {
+		regionalClient, err := dataproc.NewClusterControllerClient(ctx, option.WithEndpoint(r+"-dataproc.googleapis.com:443"))
+		if err != nil {
+			log.Errorf("[GET_CLUSTERS] Error creating dataproc client for region %s: %+v", r, err)
+			return nil, err
+		}
+		defer regionalClient.Close()
+
+		req.Region = r
+		clist, err := getClusters(regionalClient.ListClusters(ctx, req), r)
+		if err != nil {
+			log.Errorf("[GET_CLUSTERS] Error listing clusters: %+v", err)
+			return nil, err
+		}
+		if len(clist) > 0 {
+			log.Debugf("[GET_CLUSTERS] Found %d clusters in %s", len(clist), r)
+			for _, c := range clist {
```

Comment:
Are we able to use `clusters = append(clusters, clist...)` instead of the loop?

Reply:
Now that I've looked up the ellipsis syntax, I believe we are able to use it. Yes. 😅 

Thank you for pointing out a go-ism I hadn't learned yet. 😁